contributors/design-proposals/sig-cli/kubectl_apply_getsetdiff_last_applied_config.md

```diff
@@ -0,0 +1,181 @@
+# Kubectl apply subcommands for last-config
+
+## Abstract
+
+`kubectl apply` uses the `last-applied-config` annotation to compute
+the removal of fields from local object configuration files and then
+send patches to delete those fields from the live object.  Reading or
+updating the `last-applied-config` is complex as it requires parsing
+out and writing to the annotation.  Here we propose a set of porcelain
+commands for users to better understand what is going on in the system
+and make updates.
+
+## Motivation
+
+What is going on behind the scenes with `kubectl apply` is opaque.  Users
+have to interact directly with annotations on the object to view
+and make changes.  In order to stop having `apply` manage a field on
+an object, it must be manually removed from the annotation and then be removed
+from the local object configuration.  Users should be able to simply edit
+the local object configuration and set it as the last-applied-config
+to be used for the next diff base.  Storing the last-applied-config
+in an annotation adds black magic to `kubectl apply`, and it would
+help users learn and understand if the value was exposed in a discoverable
+manner.
+
+## Use Cases
+
+1. As a user, I want to be able to diff the last-applied-configuration
+against the current local configuration to see which changes the command is seeing
+2. As a user, I want to remove fields from being managed by the local
+object configuration by removing them from the local object configuration
+and setting the last-applied-configuration to match.
+3. As a user, I want to be able to view the the base
+last-applied-configuration that will be used to merge changes
+
+## Naming and Format possibilities
+
+### Naming
+
+1. last-applied
+2. last-config
+3. last-applied-config
+4. last-configuration
+5. last-applied-configuration
+6. last
+
+### Formats
+
+1. Apply subcommands
+  - `kubectl apply set/view/diff last-applied
+  - a little bit odd to have 2 verbs in a row
+  - improves discoverability to have these as subcommands so they are tied to apply
+2. Set/View subcommands
+  - `kubectl set/view/diff last-applied
+  - consistent with other set/view commands
+  - clutters discoverability of set/view commands since these are only for apply
```

Comment:
Yes that is something I have considered and had mixed feelings about.  My reasoning for organizing the commands like this is that it helps discoverability to put related commands closely together in the command tree.  As a user, I probably want to discover all the commands related to apply and last-applied-configuration in one place, e.g. kubectl apply -h should print all the sub commands related to apply.  Spreading the apply related commands across the tree will make it hard to discover new commands as well as pollute the other command namespaces.

WDYT?